cli_tools/gce_vm_image_import/importer/api_inflater.go

```diff
@@ -0,0 +1,203 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package importer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	daisyUtils "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/daisy"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/storage"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/daisycommon"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	computeAlpha "google.golang.org/api/compute/v0.alpha"
+	"google.golang.org/api/compute/v1"
+)
+
+func isCausedByUnsupportedFormat(err error) bool {
+	if err == nil {
+		return false
+	}
+	return strings.Contains(err.Error(), "INVALID_IMAGE_FILE")
+}
+
+func isCausedByAlphaAPIAccess(err error) bool {
+	if err == nil {
+		return false
+	}
+	return strings.Contains(err.Error(), "Required 'Alpha Access' permission")
+}
+
+// apiInflater implements `importer.inflater` using the Compute Engine API
+type apiInflater struct {
+	serialLogs    []string
+	args          ImportArguments
+	computeClient daisyCompute.Client
+	storageClient storage.Client
+}
+
+func createAPIInflater(args ImportArguments, computeClient daisyCompute.Client, storageClient storage.Client) inflater {
+	return apiInflater{
+		serialLogs:    []string{},
+		args:          args,
+		computeClient: computeClient,
+		storageClient: storageClient,
+	}
+}
+
+func (inflater apiInflater) traceLogs() []string {
+	return inflater.serialLogs
+}
+
+func (inflater apiInflater) addTraceLog(l string) {
+	inflater.serialLogs = append(inflater.serialLogs, l)
+}
+
+func (inflater apiInflater) inflate(ctx context.Context) (persistentDisk, error) {
+	startTime := time.Now()
+	diskName := fmt.Sprintf("shadow-disk-%v", inflater.args.ExecutionID)
+	cd := computeAlpha.Disk{
+		Name:                diskName,
+		SourceStorageObject: inflater.args.SourceFile,
+	}
+	err := inflater.computeClient.CreateDiskAlpha(inflater.args.Project, inflater.args.Zone, &cd)
+	if err != nil {
+		return persistentDisk{}, err
+	}
+
+	// Cleanup the shadow disk ignoring error
+	defer inflater.computeClient.DeleteDisk(inflater.args.Project, inflater.args.Zone, cd.Name)
+
+	bkt, objPath, err := storage.GetGCSObjectPathElements(inflater.args.SourceFile)
+	if err != nil {
+		return persistentDisk{}, err
+	}
+	sourceFile := inflater.storageClient.GetObject(bkt, objPath).GetObjectHandle()
+	attrs, err := sourceFile.Attrs(ctx)
+	if err != nil {
+		return persistentDisk{}, daisy.Errf("Failed to get source file attributes: %v", err)
+	}
+	sourceFileSizeGb := (attrs.Size-1)/1073741824 + 1
+
+	diskURI := fmt.Sprintf("zones/%s/disks/%s", inflater.args.Zone, diskName)
+	pd := persistentDisk{
+		uri:           diskURI,
+		sizeGb:        cd.SizeGb,
+		sourceGb:      sourceFileSizeGb,
```

Comment:
Can this be switched to bytes instead of GB? Due to rounding, GB is not granular enough to detect small differences.

Reply:
source file is the GCS file, it's not the inflation result. This number is just for log and won't be a part of comparison.
sizeGB is the disk size, which is always in GB granular.